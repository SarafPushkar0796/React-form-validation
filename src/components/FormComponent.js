import { Component } from 'react';
import FormErrorComponent from './FormErrorComponent';

class FormComponent extends Component{
    
    // Initial state
    constructor (props) {
        super(props);
        this.state = {
            email: '',
            password: '',

            // Added a property called formErrors which will be an object with the input field names as keys and any validation errors as their values. 
            // The initial value for each key is an empty string.
            formErrors: {email: '', password: ''},
            emailValid: false,
            passwordValid: false,
            formValid: false
        }
    }

    handleSubmit(e) {
        alert('A email address was submitted: ' + this.state.value);
        e.preventDefault();
    }

    handleUserInput = (e) => {
        const name = e.target.name;
        const value = e.target.value;
        this.setState({[name]: value}, () => { this.validateField(name, value) });
    }

    // validateField function takes fieldName as 'email' or 'password'
    validateField(fieldName, value) {
        let fieldValidationErrors = this.state.formErrors;
        let emailValid = this.state.emailValid;
        let passwordValid = this.state.passwordValid;

        switch(fieldName) {
            case 'email':
                emailValid = value.match(/^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i);
                fieldValidationErrors.email = emailValid ? '' : ' address is invalid!';
            break;
            case 'password':
                passwordValid = value.length >= 8;
                fieldValidationErrors.password = passwordValid ? '': ' is too short.';
            break;
            default: break;
        }
        this.setState(
            {
                formErrors: fieldValidationErrors,
                emailValid: emailValid,
                passwordValid: passwordValid
        }, this.validateForm);
    }

    validateForm() {
        this.setState({formValid: this.state.emailValid && this.state.passwordValid});
    }

    // render component
    render(){
        return (
            <div className="container">
                <div className="row mt-5">
                    <div className="col-lg">
                        <form className="demoForm" onSubmit={this.handleSubmit}>
                            <h2>Sign up</h2>
                            <FormErrorComponent formErrors={this.state.formErrors}/>
                            <div className="form-group mt-4">
                                <label htmlFor="email">Email address</label>

                                {/* Set value from state and handle user input onChange */}
                                <input data-testid="email" type="email" className="form-control" name="email" placeholder="Email address" value={this.state.email} onChange={this.handleUserInput}/>
                            </div>
                            <div className="form-group">
                                <label htmlFor="password">Password</label>

                                {/* Set value from state and handle user input onChange */}
                                <input data-testid="password" type="password" className="form-control" name="password" placeholder="Password" value={this.state.password} onChange={this.handleUserInput}/>
                            </div>
                            <button data-testid="button-disabled" type="submit" className="btn btn-primary" disabled={!this.state.formValid}>Sign up</button>
                        </form>
                    </div>
                </div>                
            </div>            
        )
    }
}

export default FormComponent;