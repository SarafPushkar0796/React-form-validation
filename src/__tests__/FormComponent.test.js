import { render, screen, cleanup, getByDisplayValue, fireEvent } from '@testing-library/react';
import FormComponent from '../components/FormComponent';

//  cleanup is passed as a parameter to afterEach to just clean up everything after each test to avoid memory leaks.
afterEach(cleanup);

test('renders FormComponent', () => {

    // render the FormComponent component with render and get back asFragment as a returned value
    const { asFragment } = render(<FormComponent />);    
  
    // make sure that the fragment of the FormComponent component matches the snapshot.
    expect(asFragment(<FormComponent />)).toMatchSnapshot();      
});

// email address unit test
test('check if email address is valid', () => {
    const { getByTestId } = render(<FormComponent />);

    // email check using regex
    const uriRegEx = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    
    // get email input
    const email = getByTestId('email');
    
    // element present in the document
    expect(email).toBeInTheDocument();    

    expect(email).toHaveAttribute('type','email');

    // set value in email input
    fireEvent.change(email, {
        target: {value: 'abcd@xyz.com'}
    });    

    // Check if set
    expect(email).toHaveValue('abcd@xyz.com');
    // console.log(email);

    // validate email using regex
    expect('abcd@xyz.com').toMatch(/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/);
    
    // weirdly not working
    // expect(email).toMatch(uriRegEx);
});

// password unit test
test('check if password is valid', () => {
    const { getByTestId } = render(<FormComponent />);

    // get password input
    const password = getByTestId('password');
    
    // element present in the document
    expect(password).toBeInTheDocument();    

    expect(password).toHaveAttribute('type','password');

    // set value in email input
    fireEvent.change(password, {
        target: {value: 'Hy6l0dE$sZZu'}
    });

    // Check if set
    expect(password).toHaveValue('Hy6l0dE$sZZu');
    // console.log(password);

    // validate password using regex
    /*
        Match 6 to 16 valid characters and includes a-z, A-Z, 0-9
        (?=.*[0-9]) - String has at least one number;
        (?=.*[!@#$%^&*]) - String has at least one special character.        
    */
    expect('Hy6l0dE$sZZu').toMatch(/^(?=.*[0-9])(?=.*[!@#$%^&*])[a-zA-Z0-9!@#$%^&*]{6,16}$/);
});

// check if button disabled     
test('should be disabled', () => {
    const { getByTestId } = render(<FormComponent />);

    // check attribute to be submit
    expect(getByTestId('button-disabled')).toHaveAttribute('type','submit'); 

    // check button to be disabled
    expect(getByTestId('button-disabled')).toBeDisabled()
});