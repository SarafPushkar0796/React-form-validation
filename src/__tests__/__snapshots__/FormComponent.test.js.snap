// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders FormComponent 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row mt-5"
    >
      <div
        class="col-lg"
      >
        <form
          class="demoForm"
        >
          <h2>
            Sign up
          </h2>
          <div
            class="formErrors"
          >
            <div
              class="alert alert-danger"
              role="alert"
            />
          </div>
          <div
            class="form-group mt-4"
          >
            <label
              for="email"
            >
              Email address
            </label>
            <input
              class="form-control"
              data-testid="email"
              name="email"
              placeholder="Email address"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              class="form-control"
              data-testid="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <button
            class="btn btn-primary"
            data-testid="button-disabled"
            disabled=""
            type="submit"
          >
            Sign up
          </button>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
